File: springboot-backend/src/main/java/com/app/springbackend/config/WebConfig.java
package com.app.springbackend.config;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.servlet.config.annotation.CorsRegistry;
import org.springframework.web.servlet.config.annotation.EnableWebMvc;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;

/**
 * Configures Cross-Origin Resource Sharing (CORS) for the application.
 */
@Configuration
@EnableWebMvc
public class WebConfig implements WebMvcConfigurer {

    @Value("${klv.app.frontend-server}")
    private String frontendServer;

    /**
     * Adds CORS mappings for the entire application, allows all headers and credentials.
     * The maximum age for pre-flight requests - 1800 seconds.
     *
     * @param registry the {@link CorsRegistry} to add mappings to.
     */
    @Override
    public void addCorsMappings(CorsRegistry registry) {
        registry.addMapping("/api/**")
                .allowedOriginPatterns(
                        "http://%s:3000".formatted(frontendServer),
                        "http://%s:3001".formatted(frontendServer)
                )
                .allowedMethods("GET", "POST", "PUT", "DELETE")
                .maxAge(1800)
                .allowedHeaders("*")
                .allowCredentials(true);
    }
}

File: backend/src/main/kotlin/com/app/backend/config/WebConfig.kt
package com.app.backend.config

import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Configuration
import org.springframework.web.servlet.config.annotation.CorsRegistry
import org.springframework.web.servlet.config.annotation.EnableWebMvc
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer

/**
 * Configures Cross-Origin Resource Sharing (CORS) for the application.
 *
 * @author Vladislav Nasevich
 */
@Configuration
@EnableWebMvc
class WebConfig : WebMvcConfigurer {

    @Value("\${tsr.app.frontend-port}")
    private val frontendPort: String? = null

    /**
     * Configures CORS mappings for the specified endpoints.
     *
     * @param registry [CorsRegistry] to configure.
     */
    override fun addCorsMappings(registry: CorsRegistry) {
        registry
            .addMapping("/api/**")
            .allowedOriginPatterns(
                "http://localhost:$frontendPort",
                "http://95.164.7.153:$frontendPort", // TODO: Rework
            )
            .allowedMethods("GET", "POST", "PUT", "DELETE")
            .maxAge(1800)
            .allowedHeaders("*")
            .allowCredentials(true)
    }
}
